### PR TITLE
Update ECW driver: Fix stream code & minor cosmetic changes

### PR DESCRIPTION
**Pull request by jbowman-hexagon**
## What does this PR do?
Fix stream code, minor spelling & comment fixes. Merging this PR will mean it is up to date with our repo which we can then work from in future.

## Tasklist

 - [ ] ADD YOUR TASKS HERE
 - [ ] Add test case(s)
 - [ ] Add documentation
 - [ ] Review
 - [x] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS:
* Compiler:

### CODE REVIEW

File: gdal/frmts/ecw/lookup.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # ******************************************************************************
-#  $Id$
+#  $Id: lookup.py d7e898f6034a9c6c0f83cfbff921434dfc2082be 2018-04-18 23:07:52 +1000 Ben Elliston $
```
Comment (rouault) [inline gdal/frmts/ecw/lookup.py:2]: it looks like it undoes at least part of Python cleanups that have been done. Will need more careful merging

File: gdal/frmts/ecw/ecwasyncreader.cpp [hunk 2]
```diff
@@ -6,7 +6,7 @@
  *
  ******************************************************************************
  * Copyright (c) 2011, Frank Warmerdam <warmerdam@pobox.com>
- * Copyright (c) 2013, Even Rouault <even dot rouault at spatialys.com>
+ * Copyright (c) 2013, Even Rouault <even dot rouault at mines-paris dot org>
```
Comment (rouault) [inline gdal/frmts/ecw/ecwasyncreader.cpp:9]: unwanted change

File: gdal/frmts/ecw/ecwcreatecopy.cpp [hunk 3]
```diff
@@ -6,7 +6,7 @@
  *
  ******************************************************************************
  * Copyright (c) 2001, 2004, Frank Warmerdam <warmerdam@pobox.com>
- * Copyright (c) 2008-2013, Even Rouault <even dot rouault at spatialys.com>
+ * Copyright (c) 2008-2013, Even Rouault <even dot rouault at mines-paris dot org>
```
Comment (rouault) [inline gdal/frmts/ecw/ecwcreatecopy.cpp:9]: unwanted change

File: gdal/frmts/ecw/ecwcreatecopy.cpp [hunk 4]
```diff
@@ -1625,15 +1625,15 @@ class ECWWriteDataset final: public GDALDataset
     virtual void   FlushCache( void ) override;
 
     virtual CPLErr GetGeoTransform( double * ) override;
-    virtual const char* _GetProjectionRef() override;
+    virtual const char* _GetProjectionRef() override;
```
Comment (rouault) [inline gdal/frmts/ecw/ecwcreatecopy.cpp:1628]: unwanted change

File: gdal/frmts/ecw/ecwcreatecopy.cpp [hunk 5]
```diff
@@ -1625,15 +1625,15 @@ class ECWWriteDataset final: public GDALDataset
...
+    virtual const char* _GetProjectionRef() override; 
     virtual CPLErr SetGeoTransform( double * ) override;
     virtual CPLErr _SetProjection( const char *pszWKT ) override;
     const OGRSpatialReference* GetSpatialRef() const override {
         return GetSpatialRefFromOldGetProjectionRef();
     }
     CPLErr SetSpatialRef(const OGRSpatialReference* poSRS) override {
         return OldSetProjectionFromSetSpatialRef(poSRS);
-    }
+    }
```
Comment (rouault) [inline gdal/frmts/ecw/ecwcreatecopy.cpp:1636]: unwanted change

File: gdal/frmts/ecw/ecwdataset.cpp [hunk 6]
```diff
@@ -6,7 +6,7 @@
  *
  ******************************************************************************
  * Copyright (c) 2001, Frank Warmerdam <warmerdam@pobox.com>
- * Copyright (c) 2007-2013, Even Rouault <even dot rouault at spatialys.com>
+ * Copyright (c) 2007-2013, Even Rouault <even dot rouault at mines-paris dot org>
```
Comment (rouault) [inline gdal/frmts/ecw/ecwdataset.cpp:9]: unwanted change

File: gdal/frmts/ecw/ecwdataset.cpp [hunk 7]
```diff
@@ -3555,7 +3555,7 @@ void GDALRegister_ECW()
     osLongName += ")";
 
     poDriver->SetMetadataItem( GDAL_DMD_LONGNAME, osLongName );
-    poDriver->SetMetadataItem( GDAL_DMD_HELPTOPIC, "drivers/raster/ecw.html" );
+    poDriver->SetMetadataItem( GDAL_DMD_HELPTOPIC, "frmt_ecw.html" );
```
Comment (rouault) [inline gdal/frmts/ecw/ecwdataset.cpp:3558]: unwanted change

File: gdal/frmts/ecw/ecwdataset.cpp [hunk 8]
```diff
@@ -3652,7 +3652,7 @@ void GDALRegister_JP2ECW()
     osLongName += ")";
 
     poDriver->SetMetadataItem( GDAL_DMD_LONGNAME, osLongName );
-    poDriver->SetMetadataItem( GDAL_DMD_HELPTOPIC, "drivers/raster/jp2ecw.html" );
+    poDriver->SetMetadataItem( GDAL_DMD_HELPTOPIC, "frmt_jp2ecw.html" );
```
Comment (rouault) [inline gdal/frmts/ecw/ecwdataset.cpp:3655]: unwanted change

File: gdal/frmts/ecw/ecwdataset.cpp [hunk 9]
```diff
@@ -2541,19 +2540,20 @@ CNCSJP2FileView *ECWDataset::OpenFileView( const char *pszDatasetName,
...
+       auto vsiIoStream = NCS::CView::FindSteamByStreamNameFromOpenDatasets(streamName);
+       if (!vsiIoStream) {
+           vsiIoStream = std::make_shared<VSIIOStream>();
+           std::static_pointer_cast<VSIIOStream>(vsiIoStream)->Access(fpVSIL, FALSE, TRUE, pszDatasetName, 0, -1);
+           bUsingCustomStream = TRUE;
+       }
+       oErr = poFileView->Open(vsiIoStream, bProgressive);
 #else
-        auto vsiIoStream = new VSIIOStream();
+                auto vsiIoStream = new VSIIOStream();
```
Comment (rouault) [inline gdal/frmts/ecw/ecwdataset.cpp:2556]: unwanted change

File: gdal/frmts/ecw/ecwsdk_headers.h [hunk 10]
```diff
@@ -31,6 +31,12 @@
 
 #include "cpl_port.h"
 
+#ifdef HAVE_CPL_SAFER_SNPRINTF
```
Comment (rouault) [inline gdal/frmts/ecw/ecwsdk_headers.h:34]: all changes in that file seem undesirable. those additions were removed per d1a243f0171413dfb8e9336e7f6ecde43614d185

File: gdal/frmts/ecw/frmt_ecw.html [hunk 11]
```diff
@@ -0,0 +1,224 @@
+<html>
```
Comment (rouault) [inline gdal/frmts/ecw/frmt_ecw.html:1]: Documentation no longer uses HTML files. The doc is now at gdal/doc/source/drivers/raster/ecw.rst

File: gdal/frmts/ecw/gdal_ecw.h [hunk 12]
```diff
@@ -1,13 +1,13 @@
 /******************************************************************************
- * $Id$
+ * $Id: gdal_ecw.h 7e07230bbff24eb333608de4dbd460b7312839d0 2017-12-11 19:08:47Z Even Rouault $
```
Comment (rouault) [inline gdal/frmts/ecw/gdal_ecw.h:2]: unwanted change

File: gdal/frmts/ecw/gdal_ecw.h [hunk 13]
```diff
@@ -1,13 +1,13 @@
...
  *
  * Project:  GDAL
  * Purpose:  ECW (ERDAS Wavelet Compression Format) Driver Definitions
  * Author:   Frank Warmerdam, warmerdam@pobox.com
  *
  ******************************************************************************
  * Copyright (c) 2001-2011, Frank Warmerdam <warmerdam@pobox.com>
- * Copyright (c) 2013, Even Rouault <even dot rouault at spatialys.com>
+ * Copyright (c) 2013, Even Rouault <even dot rouault at mines-paris dot org>
```
Comment (rouault) [inline gdal/frmts/ecw/gdal_ecw.h:10]: unwanted change

File: gdal/frmts/ecw/gdal_ecw.h [hunk 14]
```diff
@@ -90,8 +95,8 @@ class JP2UserBox final: public CNCSJP2Box {
     virtual ~JP2UserBox();
 
 #if ECWSDK_VERSION >= 55
-    CNCSError Parse(NCS::SDK::CFileBase &JP2File, const NCS::CIOStreamPtr &Stream) override;
-    CNCSError UnParse(NCS::SDK::CFileBase &JP2File, const NCS::CIOStreamPtr &Stream) override;
+	CNCSError Parse(NCS::SDK::CFileBase &JP2File, const NCS::CIOStreamPtr &Stream) override;
```
Comment (rouault) [inline gdal/frmts/ecw/gdal_ecw.h:98]: unwanted change

File: gdal/frmts/ecw/gdal_ecw.h [hunk 15]
```diff
@@ -90,8 +95,8 @@ class JP2UserBox final: public CNCSJP2Box {
...
+	CNCSError UnParse(NCS::SDK::CFileBase &JP2File, const NCS::CIOStreamPtr &Stream) override;
```
Comment (rouault) [inline gdal/frmts/ecw/gdal_ecw.h:99]: unwanted change

File: gdal/frmts/ecw/gdal_ecw.h [hunk 16]
```diff
@@ -120,10 +125,9 @@ class JP2UserBox final: public CNCSJP2Box {
 
 class VSIIOStream final: public CNCSJPCIOStream
 {
-#if ECWSDK_VERSION >= 54
-    NCS_DELETE_ALL_COPY_AND_MOVE(VSIIOStream)
-#endif
+	NCS_DELETE_ALL_COPY_AND_MOVE(VSIIOStream)
```
Comment (rouault) [inline gdal/frmts/ecw/gdal_ecw.h:128]: why is #if ECWSDK_VERSION >= 54 dropped ?

File: gdal/frmts/ecw/gdal_ecw.h [hunk 17]
```diff
@@ -141,7 +145,7 @@ class VSIIOStream final: public CNCSJPCIOStream
     using CNCSJPCIOStream::Read;
 
     VSIIOStream() : m_Filename(nullptr)
-    {
+	{
```
Comment (rouault) [inline gdal/frmts/ecw/gdal_ecw.h:148]: unwanted change

File: gdal/frmts/ecw/gdal_ecw.h [hunk 18]
```diff
@@ -157,7 +161,7 @@ class VSIIOStream final: public CNCSJPCIOStream
         abyCOMType[1] = 0;
     }
     virtual ~VSIIOStream() {
-        VSIIOStream::Close();
+	    VSIIOStream::Close();
```
Comment (rouault) [inline gdal/frmts/ecw/gdal_ecw.h:164]: unwanted change

File: gdal/frmts/ecw/gdal_ecw.h [hunk 19]
```diff
@@ -199,10 +203,20 @@ class VSIIOStream final: public CNCSJPCIOStream
         bSeekable = bSeekableIn;
         VSIFSeekL(fpVSIL, startOfJPData, SEEK_SET);
         m_Filename = CPLStrdup(pszFilename);
+		
+#if ECWSDK_VERSION >= 55
+		const std::string vsiStreamPrefix("STREAM=/vsi");
```
Comment (rouault) [inline gdal/frmts/ecw/gdal_ecw.h:208]: bad indentation

File: gdal/frmts/ecw/ecwasyncreader.cpp [hunk 20]
```diff
@@ -199,6 +199,11 @@ ECWAsyncReader::ECWAsyncReader()
 {
     hMutex = CPLCreateMutex();
     CPLReleaseMutex( hMutex );
+
+    poFileView = nullptr;
```
Comment (rouault) [inline gdal/frmts/ecw/ecwasyncreader.cpp:203]: unwanted change: already initialized in gdal_ecw.h

**Comment by rouault** [comment]
Another remark is that current master is already supposed to be 5.5 compatible: what does this PR exactly bring ?

File: gdal/frmts/ecw/ecwcreatecopy.cpp [hunk 21]
```diff
@@ -143,11 +143,11 @@ class GDALECWCompressor final: public CNCSFile {
 /************************************************************************/
 
 GDALECWCompressor::GDALECWCompressor() :
-    m_OStream(std::make_shared<VSIIOStream>()), eWorkDT(GDT_Unknown)
+    eWorkDT(GDT_Unknown), m_OStream(std::make_shared<VSIIOStream>())
```
Comment (rouault) [inline gdal/frmts/ecw/ecwcreatecopy.cpp:146]: Travis-CI complaining about
```
ecwcreatecopy.cpp:146:5: error: field 'eWorkDT' will be initialized after field 'm_OStream' [-Werror,-Wreorder]
    eWorkDT(GDT_Unknown), m_OStream(std::make_shared<VSIIOStream>())
    ^
```

**Comment by rouault** [comment]
@jbowman-hexagon  

Sorry to hijack this PR. Just received an email from someone who has build issues with ECW 5.5 Update 1 on Windows with GDAL 3.1.0
```
d:\jenkins_work\workspace\gdal_3.1.0\gdal-3.1.0\frmts\ecw\ecwdataset.cpp(1067): error C2440: 'type cast': cannot convert from 'NCS::CIOStreamPtr' to 'VSIIOStream *'
d:\jenkins_work\workspace\gdal_3.1.0\gdal-3.1.0\frmts\ecw\ecwdataset.cpp(1067): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
d:\jenkins_work\workspace\gdal_3.1.0\gdal-3.1.0\frmts\ecw\ecwdataset.cpp(2558): error C2664: 'NCS::CError NCS::CView::Open(const NCS::CIOStreamPtr &,bool,NCSFileType)': cannot convert argument 1 from 'VSIIOStream *' to 'wchar_t *'
d:\jenkins_work\workspace\gdal_3.1.0\gdal-3.1.0\frmts\ecw\ecwdataset.cpp(2558): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
d:\jenkins_work\workspace\gdal_3.1.0\gdal-3.1.0\frmts\ecw\ecwdataset.cpp(2575): error C2440: 'type cast': cannot convert from 'NCS::CIOStreamPtr' to 'VSIIOStream *'
d:\jenkins_work\workspace\gdal_3.1.0\gdal-3.1.0\frmts\ecw\ecwdataset.cpp(2575): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
```

 Are there been recent API changes in the SDK that would require driver changes ? (it doesn't seem like this PR would address the above compilation errors)

**Comment by jbowman-hexagon** [comment]
Hi @rouault 
There have been no API changes. Could you please file this under a separate issue and include some more details

**Pull request merged**